### ISSUE TITLE
gitlab does not support to get sub-package by ``go get``

### ISSUE THREAD

**Issue by hvnsweeting** [gitlabhq/gitlabhq#9508]
I tried this with gitlab.com, so that probably contains the newest stable version of gitlabhq:

````
$ go get -v gitlab.com/hvn_familug/pikachu/raichu
Fetching https://gitlab.com/hvn_familug/pikachu/raichu?go-get=1
Parsing meta tags from https://gitlab.com/hvn_familug/pikachu/raichu?go-get=1 (status code 200)
import "gitlab.com/hvn_familug/pikachu/raichu": parse https://gitlab.com/hvn_familug/pikachu/raichu?go-get=1: no go-import meta tags
package gitlab.com/hvn_familug/pikachu/raichu: unrecognized import path "gitlab.com/hvn_familug/pikachu/raichu"
$ go get -v gitlab.com/hvn_familug/pikachu
Fetching https://gitlab.com/hvn_familug/pikachu?go-get=1
Parsing meta tags from https://gitlab.com/hvn_familug/pikachu?go-get=1 (status code 200)
get "gitlab.com/hvn_familug/pikachu": found meta tag main.metaImport{Prefix:"gitlab.com/hvn_familug/pikachu", VCS:"git", RepoRoot:"https://gitlab.com/hvn_familug/pikachu.git"} at https://gitlab.com/hvn_familug/pikachu?go-get=1
````

On my gitlab 7.3 private server, it's the same.

When github, and golang.org work, they all returned correct RepoRoot:

````
$ go get -v github.com/astaxie/beego/config
github.com/astaxie/beego (download)

$ godep restore -v
...
 godep: get "golang.org/x/crypto/cast5": found meta tag vcs.metaImport{Prefix:"golang.org/x/crypto", VCS:"git", RepoRoot:"https://go.googlesource.com/crypto"} at https://golang.org/x/crypto/cast5$
go-get=1
godep: get "golang.org/x/crypto/cast5": verifying non-authoritative meta tag
godep: get "golang.org/x/crypto/openpgp": found meta tag vcs.metaImport{Prefix:"golang.org/x/crypto", VCS:"git", RepoRoot:"https://go.googlesource.com/crypto"} at https://golang.org/x/crypto/ope$
pgp?go-get=1
````

https://github.com/gitlabhq/gitlabhq/pull/5958 only support get the package if it's at root of the repo.

**Comment by hvnsweeting** [comment]
hi, anybody here? :hand:

**Comment by stanhu** [comment]
Yeah, this doesn't work on GitLab at the moment. Support for Go retrieval was added in #5958 and #7693, but unfortunately to support sub-packages I think something in GitLab will have to be changed. At the moment, attempting to retrieve a sub-package URL as the follows, results in a 302:

```
$ curl https://gitlab.com/gitlab-org/gitlab-ce/lib?go-get=1
<html><body>You are being <a href="https://gitlab.com/users/sign_in">redirected</a>.</body></html>
```

I could see a case where a URL with `go-get=1` query parameter would be routed to a controller to check for the package/directory existence and permissions, and then return the meta tags if those checks pass.

**Comment by naxhh** [comment]
+1
Go is our principal use case for gitlab at the moment.

**Comment by ethernetdan** [comment]
+1
This issue is currently forcing us to move off Gitlab. Any suggestions on how it could be implemented? I'd be interested in putting in a PR.

**Comment by ingardm** [comment]
any traction on this?

**Comment by stanhu** [comment]
@ethernetdan If you're interested in doing this, you may want to look at adding something to `projects_controller.rb`. You might be able to create a route with a wildcard matching route (since anything after the project part of the path could be part of the Go subpackage) and check if the `go-get` parameter is there.

**Comment by dimchansky** [comment]
+1

**Comment by Lswith** [comment]
+1

**Comment by adamchalmers** [comment]
+1

**Comment by porkcharsui** [comment]
Since I'm using NGINX, under config `/var/opt/gitlab/nginx/conf/gitlab-http.conf` I added this to fix my build woes:
```
if ($args ~* "^go-get=1") {
  set $condition goget;
}
if ($uri ~ ^/([a-zA-Z0-9_-]+)/([a-zA-Z0-9_-]+)/.*$) {
  set $condition "${condition}path";   
}
if ($condition = gogetpath) {
  return 200 "<!DOCTYPE html><html><head><meta content='your.domain.com/$1/$2 git http://your.domain.com/$1/$2.git' name='go-import'></head></html>";
}

```
Change the FQDN to fit your setup.

**Comment by leetal** [comment]
A workaround like this needs to be implemented into Gitlab asap ;) We are also relying on gitlab and we really don't want to move to another repo-provider.

**Comment by stanhu** [comment]
@JobV, wanted to get this on your radar. It doesn't sound like that much work, but it might be a bit trickier to get right than I think at the moment.

**Comment by porkcharsui** [comment]
@stanhu The meta tag implementation is pretty straight forward and simply tells `go get` where it should git clone from and for Gitlab this is always `group/project` for any package under the project.

The only issues I have encountered testing for a workaround occur when this code path is triggered improperly for paths like these: 
* group/project/
* group/project.git/
(Note: I have updated the regex in my original post to fix for this)

**Comment by Lswith** [comment]
here is working code that should fix this problem in the nginx config:

    if ($args ~* "^go-get=1") { 
        set $condition goget; 
    }
    if ($uri ~ "^/([\w-]+)/([\w-]+)/[/\w-]+$") { 
        set $condition "${condition}path"; 
    }
    if ($condition = gogetpath) {
         add_header Content-Type text/html;
         return 200 "<!DOCTYPE html><html><head><meta content='your.domain.com/$1/$2 git https://your.domain.com/$1/$2.git' name='go-import'></head></html>";
    }

**Comment by DouweM** [comment]
Thank you for your suggestions everyone! I just had a little brainstorm session about this with @jacobvosmaer, and the best idea we came up with is to use a Rake middleware to intercept any `?go_get=1` requests and return a very basic document with the expected meta-tag based on the request path. 

This keeps the logic close to the Rails app (rather than out in nginx config or gitlab-workhorse), but doesn't affect regular Rails routing, which doesn't currently support the `<group>/<project>/<dir>` format, instead expecting `<group>/<project>/tree/<dir>`.

@Lswith How does this sound?

**Comment by jacobvosmaer** [comment]
I thought about it some more. One downside of catching `?go_get=1` in a middleware layer is that it will return bogus results instead of a 404 if a user requests a resource that does not exist. That is not nice, and using config/routes.rb and a controller would allow us to give more authoritative answers.

I am not 100% sure that there is no security angle to referring to potentially non-existent repos from a middleware. @DouweM do you see any sort of repo squatting possibilities?

**Comment by stanhu** [comment]
@jacobvosmaer Right, that's why I think it should be in a controller. The user also has to have access to the repo even if it exists.

**Comment by DouweM** [comment]
@jacobvosmaer `go get` doesn't have authentication, so if we want to support private repos, we _need_ to ignore whether the repo actually exists.

**Comment by jacobvosmaer** [comment]
@DouweM @stanhu yes if the implementation is dumb (middleware) it cannot
leak information, maybe that is best after all.

**Comment by stanhu** [comment]
@DouweM @jacobvosmaer :+1 for dumb implementations that work well--that's counterintuitive. :)

**Comment by sbinet** [comment]
+1
this is also blocking publishing documentation via `godoc`: golang/gddo#356

**Comment by jacobvosmaer** [comment]
I created an issue on the official GitLab issue tracker for this so we can give it a milestone. https://gitlab.com/gitlab-org/gitlab-ce/issues/13805

**Comment by stanhu** [comment]
This should work now via e23d6ada and should be available in GitLab 8.6.

**Comment by jacobvosmaer** [comment]
Thanks @stanhu !
